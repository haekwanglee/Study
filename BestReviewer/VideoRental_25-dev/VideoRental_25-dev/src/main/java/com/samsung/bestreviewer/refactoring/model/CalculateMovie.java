package com.samsung.bestreviewer.refactoring.model;

public abstract class CalculateMovie {

    public static CalculateMovie createByType(int priceCode) {
        switch (priceCode) {
            case Movie.REGULAR:
                return new CalculateMovieRegular();
            case Movie.NEW_RELEASE:
                return new CalculateMovieNewRelease();
            case Movie.CHILDREN:
                return new CalculateMovieChildren();
            case Movie.BEST_SELLER:
                return new CalculateMovieBestseller();
            default:
                return new CalculateMovieDefault();
        }
    }

    public abstract double getAmount(int daysRented);
    public abstract int getFrequentRenterPoints(int daysRented);
    public abstract int getFrequentBonusPoints(int daysRented);
}