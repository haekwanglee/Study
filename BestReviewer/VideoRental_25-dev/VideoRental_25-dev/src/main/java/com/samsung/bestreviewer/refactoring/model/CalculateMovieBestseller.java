package com.samsung.bestreviewer.refactoring.model;

public class CalculateMovieBestseller extends CalculateMovie {
    @Override
    public double getAmount(int daysRented) {
        double amount = 0;
        if (daysRented <= 2) {
            amount += daysRented * 5;
        } else {
            amount += 2 * 5 + (daysRented - 2) * 7.5;
        }
        return amount;
    }

    @Override
    public int getFrequentRenterPoints(int daysRented) {
        return daysRented * 5;
    }

    @Override
    public int getFrequentBonusPoints(int daysRented) {
        return 0;
    }

}
