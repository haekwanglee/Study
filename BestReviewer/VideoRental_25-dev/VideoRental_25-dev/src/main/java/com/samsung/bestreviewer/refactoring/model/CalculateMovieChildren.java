package com.samsung.bestreviewer.refactoring.model;

public class CalculateMovieChildren extends CalculateMovie {
    @Override
    public double getAmount(int daysRented) {
        double amount = 0;
        amount += 1.5;
        if (daysRented > 3)
            amount += (daysRented - 3) * 1.5;
        return amount;
    }

    @Override
    public int getFrequentRenterPoints(int daysRented) {
        return 1;
    }

    @Override
    public int getFrequentBonusPoints(int daysRented) {
        return 0;
    }
}
