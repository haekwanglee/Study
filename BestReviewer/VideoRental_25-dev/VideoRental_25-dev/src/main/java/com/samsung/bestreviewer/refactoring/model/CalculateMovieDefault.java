package com.samsung.bestreviewer.refactoring.model;

public class CalculateMovieDefault extends CalculateMovie {
    @Override
    public double getAmount(int daysRented) {
        return 0;
    }

    @Override
    public int getFrequentRenterPoints(int daysRented) {
        return 0;
    }

    @Override
    public int getFrequentBonusPoints(int daysRented) {
        return 0;
    }
}
