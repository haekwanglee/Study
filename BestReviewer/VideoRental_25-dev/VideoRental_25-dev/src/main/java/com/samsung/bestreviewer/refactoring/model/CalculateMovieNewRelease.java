package com.samsung.bestreviewer.refactoring.model;

public class CalculateMovieNewRelease extends CalculateMovie {
    @Override
    public double getAmount(int daysRented) {
        double amount = 0;
        amount += daysRented * 3;
        return amount;
    }

    @Override
    public int getFrequentRenterPoints(int daysRented) {
        return 1;
    }

    @Override
    public int getFrequentBonusPoints(int daysRented) {
        if (daysRented > 1) {
            return 1;
        }
        return 0;
    }
}
