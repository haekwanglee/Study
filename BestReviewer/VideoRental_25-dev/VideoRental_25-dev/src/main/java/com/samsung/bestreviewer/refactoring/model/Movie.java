package com.samsung.bestreviewer.refactoring.model;

public class Movie {

    public static final int CHILDREN = 2;
    public static final int REGULAR = 0;
    public static final int NEW_RELEASE = 1;
    public static final int BEST_SELLER = 3;

    private String title;
    private int priceCode;
    private CalculateMovie calculateMovie;

    public Movie(String title, int priceCode) {
        this.title = title;
        this.priceCode = priceCode;
        this.calculateMovie = CalculateMovie.createByType(priceCode);
    }

    public int getPriceCode() {
        return priceCode;
    }

    public String getTitle() {
        return title;
    }

    public CalculateMovie getCalculateMovie() {
        return calculateMovie;
    }
}
