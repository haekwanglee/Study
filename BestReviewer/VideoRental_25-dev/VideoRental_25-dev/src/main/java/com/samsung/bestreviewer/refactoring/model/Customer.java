package com.samsung.bestreviewer.refactoring.model;

import com.samsung.bestreviewer.refactoring.statement.StatementData;

import java.util.Enumeration;
import java.util.Vector;

public class Customer {
    private String name;
    private Vector rentals = new Vector();
    private int frequentRenterPoints = 0;

    public Customer(String name) {
        this.name = name;
        frequentRenterPoints = 0;
    }

    public void addRental(Rental rental) {
        rentals.add(rental);
    }

    public String getName() {
        return name;
    }

    public Vector getRental() {
        return rentals;
    }

    public int getFrequentRenterPoints() {
        return frequentRenterPoints;
    }

    public void addFrequentRenterPoints(int frequentRenterPoints) {
        this.frequentRenterPoints += frequentRenterPoints;
    }

}

