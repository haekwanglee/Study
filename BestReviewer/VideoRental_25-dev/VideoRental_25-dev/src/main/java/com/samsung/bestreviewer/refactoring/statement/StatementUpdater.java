package com.samsung.bestreviewer.refactoring.statement;

import com.samsung.bestreviewer.refactoring.model.Customer;
import com.samsung.bestreviewer.refactoring.model.Rental;

public class StatementUpdater {

    private Customer customer;
    private StatementData statementData;
    private StatementPrinter statementPrinterOrigin;
    private StatementPrinter statementPrinterHtml;

    public StatementUpdater(Customer customer) {
        this.customer = customer;
        statementData = new StatementData(customer.getRental());
        statementPrinterOrigin = new StatementPrinterOrigin(this);
        statementPrinterHtml = new StatementPrinterHtml(this);
    }

    public Customer getCustomer() {
        return customer;
    }

    public StatementData getStatementData() {
        return statementData;
    }

    public StatementPrinter getStatementPrinterOrigin() {
        return statementPrinterOrigin;
    }

    public StatementPrinter getStatementPrinterHtml() {
        return statementPrinterHtml;
    }

    public void updateStatement() {
        initStatementData();

        printHeader();

        for (Rental rental: statementData.getRentals()) {
            initCurrentAmount();
            updateAmountAndPoints(rental);
            printFigures(rental);
        }

        printFooter();

        customer.addFrequentRenterPoints(statementData.getFrequentRenterPoints());
    }

    private void printFooter() {
        statementPrinterOrigin.printFooter();
        statementPrinterHtml.printFooter();
    }

    private void printFigures(Rental rental) {
        statementPrinterOrigin.printFigures(rental);
        statementPrinterHtml.printFigures(rental);
    }

    private void printHeader() {
        statementPrinterOrigin.printHeader();
        statementPrinterHtml.printHeader();
    }

    private void initStatementData() {
        statementData.init();
    }

    private void initCurrentAmount() {
        statementData.setCurrentAmount(0);
    }

    private void updateAmountAndPoints(Rental rental) {
        updateCurrentAmount(rental);
        updateFrequentRenterPoints(rental);
        updateTotalAmount();
    }

    private void updateCurrentAmount(Rental rental) {
        double currentAmount = statementData.getCurrentAmount();
        int daysRented = rental.getDaysRented();
        currentAmount += rental.getMovie().getCalculateMovie().getAmount(daysRented);

        statementData.setCurrentAmount(currentAmount);
    }

    private void updateFrequentRenterPoints(Rental rental) {
        int daysRented = rental.getDaysRented();
        int frequentRenterPoints = statementData.getFrequentRenterPoints();
        frequentRenterPoints += rental.getMovie().getCalculateMovie().getFrequentRenterPoints(daysRented);
        frequentRenterPoints += rental.getMovie().getCalculateMovie().getFrequentBonusPoints(daysRented);

        statementData.setFrequentRenterPoints(frequentRenterPoints);
    }

    private void updateTotalAmount() {
        double totalAmount = statementData.getTotalAmount();
        statementData.setTotalAmount(totalAmount + statementData.getCurrentAmount());
    }

}

