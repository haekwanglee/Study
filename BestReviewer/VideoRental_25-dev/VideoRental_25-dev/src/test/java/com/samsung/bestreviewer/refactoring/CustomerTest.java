package com.samsung.bestreviewer.refactoring;

import com.samsung.bestreviewer.refactoring.model.Customer;
import com.samsung.bestreviewer.refactoring.model.Movie;
import com.samsung.bestreviewer.refactoring.model.Rental;
import com.samsung.bestreviewer.refactoring.statement.StatementUpdater;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

public class CustomerTest {

    @Test
    @DisplayName("Test using approvals.verify()")
    public void goldenMasterTest() {
        Customer testCustomer = new Customer("test_customer");
        makeRental(testCustomer);

        StatementUpdater statementUpdater = new StatementUpdater(testCustomer);
        statementUpdater.updateStatement();
        Approvals.verify(statementUpdater.getStatementPrinterOrigin().getPrintResult());
    }

    @Test
    @DisplayName("Test using approvals.verify() :HTML")
    public void goldenMasterTestHtml() {
        Customer testCustomer = new Customer("test_customer");
        makeRental(testCustomer);

        StatementUpdater statementUpdater = new StatementUpdater(testCustomer);
        statementUpdater.updateStatement();
        Approvals.verify(statementUpdater.getStatementPrinterHtml().getPrintResult());
    }

    private void makeRental(Customer customer) {
        List<List<Object>> goldenMasterTestParams = Arrays.asList(
                Arrays.asList("test_title", Movie.REGULAR, 1),
                Arrays.asList("test_title", Movie.REGULAR, 2),
                Arrays.asList("test_title", Movie.REGULAR, 3),
                Arrays.asList("test_title", Movie.REGULAR, 4),
                Arrays.asList("test_title", Movie.CHILDREN, 1),
                Arrays.asList("test_title", Movie.CHILDREN, 2),
                Arrays.asList("test_title", Movie.CHILDREN, 3),
                Arrays.asList("test_title", Movie.CHILDREN, 4),
                Arrays.asList("test_title", Movie.NEW_RELEASE, 1),
                Arrays.asList("test_title", Movie.NEW_RELEASE, 2),
                Arrays.asList("test_title", Movie.NEW_RELEASE, 3),
                Arrays.asList("test_title", Movie.NEW_RELEASE, 4),
                Arrays.asList("test_title", Movie.BEST_SELLER, 1),
                Arrays.asList("test_title", Movie.BEST_SELLER, 2),
                Arrays.asList("test_title", Movie.BEST_SELLER, 3),
                Arrays.asList("test_title", Movie.BEST_SELLER, 4)
        );

        Movie testMovie = null;
        Rental testRental = null;
        for (List<Object> testParamList: goldenMasterTestParams) {
            testMovie = new Movie((String)testParamList.get(0), (int)testParamList.get(1));
            testRental = new Rental(testMovie, (int)testParamList.get(2));
            customer.addRental(testRental);
        }
        return;
    }
}
