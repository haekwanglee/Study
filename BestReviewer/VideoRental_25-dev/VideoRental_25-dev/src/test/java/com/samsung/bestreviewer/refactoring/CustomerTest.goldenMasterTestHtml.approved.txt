<table><tr><th colspan=3 align=left><b>Rentals for <b></th><th><em>test_customer</em></th></tr>
<tr><td> </td><td>test_title</td><td> </td><td>2.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>2.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>3.5</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>5.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>1.5</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>1.5</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>1.5</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>3.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>3.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>6.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>9.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>12.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>5.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>10.0</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>17.5</td></tr>
<tr><td> </td><td>test_title</td><td> </td><td>25.0</td></tr>
<tr><td colspan=4><b>Amount owed is <EM>107.5</b></td></tr>
<tr><td colspan=4><b>You earned <EM>65</EM> frequent renter points</b></td></tr></table>